sw/airborne/subsystems/stereoprotocol.c

```diff
@@ -0,0 +1,167 @@
+/*
+ * stereoprotocol.cpp
+ *
+ *  Created on: Sep 23, 2015
+ *      Author: roland
+ */
+
+#include "stereoprotocol.h"
+
+
+/**
+ * Increment circular buffer counter by i
+ */
+uint16_t stereoprot_add(uint16_t counter, uint16_t i, uint16_t buffer_size)
+{
+  return (counter + i) % buffer_size;
+}
+
+
+/**
+ * Decrement circular buffer counter by i
+ */
+uint16_t stereoprot_diff(uint16_t counter, uint16_t i,uint16_t buffer_size)
+{
+  return (counter - i + buffer_size) % buffer_size;
+}
+
+/**
+ * Checks if the sequence in the array is equal to 255-0-0-171,
+ * as this means that this is the end of an image
+ */
+uint8_t stereoprot_isEndOfMsg(uint8_t *stack, uint16_t i,uint16_t buffer_size)
+{
+
+  if (stack[i] == 255 && (stack[stereoprot_add(i, 1,buffer_size)] == 0) && (stack[stereoprot_add(i, 2,buffer_size)] == 0) && stack[stereoprot_add(i, 3,buffer_size)] == 171) {
+	  return 1;
+  }
+  return 0;
+}
+
+
+
+/**
+ * Checks if the sequence in the array is equal to 255-0-0-171,
+ * as this means a new image is starting from here
+ */
+uint8_t stereoprot_isStartOfMsg(uint8_t *stack, uint16_t i,uint16_t buffer_size)
+{
+	//printf("Checking start: %d %d %d %d \n",stack[i],stack[stereoprot_add(i, 1,buffer_size)],stack[stereoprot_add(i, 2,buffer_size)],stack[stereoprot_add(i, 3,buffer_size)]);
+  if (stack[i] == 255 && (stack[stereoprot_add(i, 1,buffer_size)] == 0) && (stack[stereoprot_add(i, 2,buffer_size)] == 0) && stack[stereoprot_add(i, 3,buffer_size)] == 175) {
+    return 1;
+  }
+  return 0;
+}
+
+//void stereoprot_get_msg_properties(uint8_t *, MsgProperties *, uint16_t,uint16_t);
+
+
+void WritePart(struct link_device *dev,uint8_t* code,uint8_t length){
+
+    for(uint8_t index=0; index < length; index++){
+        uint8_t toWrite = code[index];
+        dev->put_byte(dev->periph,code[index]);
```

Comment:
You should check that there is space in the transmit buffer before trying to write to it with 

dev->check_free_space(struct uart_periph *p, uint8_t len);

Currently in the uart_arch ofr stm32 if the buffer is full it simply returns and doesn't give any indication that there is something wrong.

void uart_put_byte(struct uart_periph *p, uint8_t data)
{
  uint16_t temp = (p->tx_insert_idx + 1) % UART_TX_BUFFER_SIZE;

  if (temp == p->tx_extract_idx) {
    return;  // no room
  }
...